REGRESSION(256008@main) TestWebKitAPI.IPCTestingAPI.CanIntercept* tests fail

kkinnunen-apple opened the pull request:
#### 7e3be0d6957422b38449d50b19a4f05ab2151f5a
<pre>
REGRESSION(256008@main) TestWebKitAPI.IPCTestingAPI.CanIntercept* tests fail
<a href="https://bugs.webkit.org/show_bug.cgi?id=247256">https://bugs.webkit.org/show_bug.cgi?id=247256</a>

rdar://problem/101749810

Reviewed by Ryosuke Niwa.

The commit 256008@main changed the encoding for messages that have async
replies:
 - Before, the reply id was encoded as the first parameter.
 - After, the reply id was encoded as the last parameter.

This broke the JS IPC tester API where message listener code was not
adjusted accordingly. This caused the arguments to be decoded incorrectly.

Fix the JS IPC tester API.

Add a test testing that listener API can be used to obtain the async reply
id.

* LayoutTests/ipc/message-listener-async-message-reply-id-expected.txt: Added.
* LayoutTests/ipc/message-listener-async-message-reply-id.html: Added.
 Source/WebKit/CMakeLists.txt:
* Source/WebKit/DerivedSources-input.xcfilelist:
* Source/WebKit/DerivedSources-output.xcfilelist:
* Source/WebKit/DerivedSources.make:
* Source/WebKit/Shared/IPCTester.cpp:
(WebKit::IPCTester::sendAsyncMessageToProxy):
* Source/WebKit/Shared/IPCTester.h:
* Source/WebKit/Shared/IPCTester.messages.in:
* Source/WebKit/Shared/IPCTesterReceiver.cpp: Copied from Source/WebKit/WebProcess/GPU/graphics/ThreadSafeRemoteImageBufferFlusher.cpp.
(WebKit::IPCTesterReceiver::asyncMessage):
* Source/WebKit/Shared/IPCTesterReceiver.h: Copied from Source/WebKit/WebProcess/GPU/graphics/ThreadSafeRemoteImageBufferFlusher.cpp.
* Source/WebKit/Shared/IPCTesterReceiver.messages.in: Copied from Source/WebKit/Shared/IPCTester.messages.in.
* Source/WebKit/Sources.txt:
* Source/WebKit/WebKit.xcodeproj/project.pbxproj:
* Source/WebKit/WebProcess/WebPage/IPCTestingAPI.cpp:
(WebKit::IPCTestingAPI::JSIPC::staticFunctions):
(WebKit::IPCTestingAPI::JSIPC::addTesterReceiver):
(WebKit::IPCTestingAPI::JSIPC::removeTesterReceiver):
(WebKit::IPCTestingAPI::JSMessageListener::jsDescriptionFromDecoder):

Canonical link: <a href="https://commits.webkit.org/256420@main">https://commits.webkit.org/256420@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/6c1ea558a4366733228f97362a3350342634ccf6

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/95659 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/4921 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/28708 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/105238 "Built successfully") | [✅ 🛠 🧪 win](https://ews-build.webkit.org/#/builders/10/builds/165538 "Built successfully and passed tests") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/4989 "Built successfully") | [✅ 🛠 mac-debug](https://ews-build.webkit.org/#/builders/71/builds/33676 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/88036 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/101086 "Built successfully") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/101320 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/78/builds/3662 "Passed tests") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/82278 "Built successfully") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/30721 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/9/builds/85528 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/3/builds/87437 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/34/builds/73552 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/66/builds/39409 "Built successfully") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/73/builds/18978 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/67/builds/37108 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/20291 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/4440 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/63/builds/41104 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/60/builds/42941 "Passed tests") | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/65/builds/43091 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/39542 "Passed tests") | | 
<!--EWS-Status-Bubble-End-->

rniwa commented on Source/WebKit/Shared/IPCTesterProxy.h:
```diff
@@ -0,0 +1,53 @@
+    ~IPCTesterProxy() = default;
+
+    // IPC::MessageReceiver overrides.
+    void didReceiveMessage(IPC::Connection&, IPC::Decoder&);
```
We should use final keyword instead of adding a comment like this.

cdumez replied:
I 100% sure we should mark overrides as `final` whenever possible (or `override` otherwise).

With regards to the comment though, I think this is pretty common style in WebKit2 to indicate which base class the overrides are for, though we usually use `// IPC::MessageReceiver` instead of `// IPC::MessageReceiver overrides.`.

kkinnunen-apple replied:
Yeah, it was missing `final` by accident. Fixed, thanks.
I typically have used "// ... overrides." as hand wavy arguably that is allowed by the code style where the suggested comment is not.
Opened a https://bugs.webkit.org/show_bug.cgi?id=247313 to allow and explain the comment requested here.

rniwa commented on Source/WebKit/Shared/IPCTesterProxy.h:
```diff
@@ -0,0 +1,53 @@
+    void didReceiveMessage(IPC::Connection&, IPC::Decoder&);
+private:
+    // Messages.
+    void asyncMessage(uint32_t, CompletionHandler<void(uint32_t)>&&);
```
Ditto.

rniwa commented on Source/WebKit/Shared/IPCTesterProxy.h:
```diff
@@ -0,0 +1,53 @@
+namespace WebKit {
+
+// Proxy interface to test various IPC activities.
+class IPCTesterProxy final : public IPC::MessageReceiver {
```
I would call this IPCTestingReceiver or something.
"Proxy" suffix is rather confusing because it's not a proxy to anything.

rniwa commented on Source/WebKit/Shared/IPCTesterProxy.cpp:
```diff
@@ -0,0 +1,37 @@
+ * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF
+ * THE POSSIBILITY OF SUCH DAMAGE.
+ */
+#include "config.h"
```
Nit: Need a blank line here between */ and #include.

rniwa commented on Source/WebKit/Shared/IPCTesterProxy.cpp:
```diff
@@ -0,0 +1,37 @@
+#include "IPCTesterProxy.h"
+
+#if ENABLE(IPC_TESTING_API)
+namespace WebKit {
```
Also need a blank line between #if and namespace.

kkinnunen-apple replied:
Done. Opened https://bugs.webkit.org/show_bug.cgi?id=247314 to fix these in style rules.

rniwa commented on LayoutTests/ipc/message-listener-async-message-reply-id.html:
```diff
@@ -0,0 +1,38 @@
+    }
+}, "Intercepting async messages contains all the expected data.");
+</script>
+</body>
```
Nit: Add a new line at the end?

kkinnunen-apple replied:
Done.
I assume by rule "test are source and source typically should end in newline so that sources are regular or old patch and diff programs are able to process them or both".
I'll open a bug about style rule for ending test .htmls, .css, .js with new line.

rniwa commented on LayoutTests/ipc/message-listener-async-message-reply-id.html:
```diff
@@ -0,0 +1,38 @@
+        return;
+    const testerID = 447;
+    const processTarget = 'UI'; // Currently only UI is supported for addTesterProxy.
+    // Test starts here.
```
I don't think this comment is useful. Remove?

rniwa reviewed (approved):
r=me provided all my comments are addressed.

kkinnunen-apple committed:
REGRESSION(256008@main) TestWebKitAPI.IPCTestingAPI.CanIntercept* tests fail
https://bugs.webkit.org/show_bug.cgi?id=247256

rdar://problem/101749810

Reviewed by Ryosuke Niwa.

The commit 256008@main changed the encoding for messages that have async
replies:
 - Before, the reply id was encoded as the first parameter.
 - After, the reply id was encoded as the last parameter.

This broke the JS IPC tester API where message listener code was not
adjusted accordingly. This caused the arguments to be decoded incorrectly.

Fix the JS IPC tester API.

Add a test testing that listener API can be used to obtain the async reply
id.

* LayoutTests/ipc/message-listener-async-message-reply-id-expected.txt: Added.
* LayoutTests/ipc/message-listener-async-message-reply-id.html: Added.
 Source/WebKit/CMakeLists.txt:
* Source/WebKit/DerivedSources-input.xcfilelist:
* Source/WebKit/DerivedSources-output.xcfilelist:
* Source/WebKit/DerivedSources.make:
* Source/WebKit/Shared/IPCTester.cpp:
(WebKit::IPCTester::sendAsyncMessageToProxy):
* Source/WebKit/Shared/IPCTester.h:
* Source/WebKit/Shared/IPCTester.messages.in:
* Source/WebKit/Shared/IPCTesterReceiver.cpp: Copied from Source/WebKit/WebProcess/GPU/graphics/ThreadSafeRemoteImageBufferFlusher.cpp.
(WebKit::IPCTesterReceiver::asyncMessage):
* Source/WebKit/Shared/IPCTesterReceiver.h: Copied from Source/WebKit/WebProcess/GPU/graphics/ThreadSafeRemoteImageBufferFlusher.cpp.
* Source/WebKit/Shared/IPCTesterReceiver.messages.in: Copied from Source/WebKit/Shared/IPCTester.messages.in.
* Source/WebKit/Sources.txt:
* Source/WebKit/WebKit.xcodeproj/project.pbxproj:
* Source/WebKit/WebProcess/WebPage/IPCTestingAPI.cpp:
(WebKit::IPCTestingAPI::JSIPC::staticFunctions):
(WebKit::IPCTestingAPI::JSIPC::addTesterReceiver):
(WebKit::IPCTestingAPI::JSIPC::removeTesterReceiver):
(WebKit::IPCTestingAPI::JSMessageListener::jsDescriptionFromDecoder):

Canonical link: https://commits.webkit.org/256420@main
